File: manual/code/generate_strategies/RA(name)/cpp/cpp/function_statsmodels_perform_ols_regression.cpp
#include <string>
#include <sstream>
#include <stdexcept>
#include <vector>
#include <algorithm>
#include <numeric>
#include <Eigen/Dense>
#include <nlohmann/json.hpp> 
#include <cctype>
#include <iterator>

using json = nlohmann::json;

class OLSRegression {
public:
    std::string perform_ols_regression(const std::string &data_input, const std::string &dependent_var = "", const std::string &independent_vars_str = "") {
        json data;
        try {
            data = json::parse(data_input);
        }
        catch (...) {
            throw std::invalid_argument("Invalid JSON format.");
        }
        if (!data.is_object()) {
            throw std::invalid_argument("Data must be a dictionary.");
        }
        if (data.find(dependent_var) == data.end()) {
            throw std::out_of_range("Dependent variable not found.");
        }
        std::vector<double> y_vec;
        try {
            for (const auto &val : data[dependent_var]) {
                if (!val.is_number()) {
                    throw std::invalid_argument("Dependent variable contains non-numeric data.");
                }
                y_vec.push_back(val.get<double>());
            }
        }
        catch (...) {
            throw std::invalid_argument("Dependent variable contains non-numeric data.");
        }
        if (y_vec.size() < 2) {
            throw std::invalid_argument("Insufficient data for regression.");
        }
        int n = static_cast<int>(y_vec.size());
    
        std::vector<std::string> indep_vars;
        {
            std::istringstream iss(independent_vars_str);
            std::string token;
            while (std::getline(iss, token, ',')) {
                token.erase(token.begin(), std::find_if(token.begin(), token.end(), [](unsigned char ch) {
                    return !std::isspace(ch);
                }));
                token.erase(std::find_if(token.rbegin(), token.rend(), [](unsigned char ch) {
                    return !std::isspace(ch);
                }).base(), token.end());
                if (!token.empty()) {
                    indep_vars.push_back(token);
                }
            }
        }
    
        std::vector<std::vector<double>> Xcols;
        for (const auto &var : indep_vars) {
            if (data.find(var) == data.end()) {
                throw std::out_of_range("Independent variable " + var + " not found.");
            }
            std::vector<double> col;
            try {
                for (const auto &val : data[var]) {
                    if (!val.is_number())
                        throw std::invalid_argument("Independent variable contains non-numeric data.");
                    col.push_back(val.get<double>());
                }
            }
            catch (...) {
                throw std::invalid_argument("Independent variable contains non-numeric data.");
            }
            if (col.size() != static_cast<size_t>(n)) {
                throw std::invalid_argument("Inconsistent data length among variables.");
            }
            Xcols.push_back(col);
        }
        int p = static_cast<int>(Xcols.size());
    
        Eigen::MatrixXd X(n, p + 1);
        X.col(0) = Eigen::VectorXd::Ones(n);
        for (int j = 0; j < p; j++) {
            for (int i = 0; i < n; i++) {
                X(i, j + 1) = Xcols[j][i];
            }
        }
        Eigen::VectorXd y(n);
        for (int i = 0; i < n; i++) {
            y(i) = y_vec[i];
        }
    
        Eigen::VectorXd beta = (X.transpose() * X).ldlt().solve(X.transpose() * y);
        Eigen::VectorXd y_hat = X * beta;
        double ss_res = (y - y_hat).squaredNorm();
        double y_mean = y.mean();
        double ss_tot = (y.array() - y_mean).square().sum();
        double r_squared = 1 - ss_res / ss_tot;
    
        std::ostringstream coeffStream;
        coeffStream << "{";
        coeffStream << "\"const\": " << beta(0);
        for (int j = 0; j < p; j++) {
            coeffStream << ", \"" << indep_vars[j] << "\": " << beta(j + 1);
        }
        coeffStream << "}";
    
        std::ostringstream oss;
        oss << "R-squared: " << r_squared << ", Coefficients: " << coeffStream.str();
        return oss.str();
    }
};

File: manual/code/generate_strategies/RA(name)/java/java/function_statsmodels_perform_ols_regression.java
package com.example;

import java.util.*;
import java.util.regex.*;
import org.apache.commons.math3.stat.regression.OLSMultipleLinearRegression;

public class OLSRegression {

    // 所有逻辑均在此唯一方法内实现
    public String perform_ols_regression(String data_str, String dependent_var, String independent_vars_str) {
        // 将单引号替换为双引号，便于使用正则解析
        String json = data_str.replace("'", "\"");

        // 使用正则解析字典，匹配 "key": [value1, value2, ...]
        Map<String, List<Double>> dataMap = new HashMap<>();
        Pattern pattern = Pattern.compile("\"(.*?)\"\\s*:\\s*\\[(.*?)\\]");
        Matcher matcher = pattern.matcher(json);
        while(matcher.find()){
            String key = matcher.group(1);
            String valuesStr = matcher.group(2).trim();
            List<Double> values = new ArrayList<>();
            if(!valuesStr.isEmpty()){
                String[] parts = valuesStr.split(",");
                for(String part: parts){
                    try{
                        values.add(Double.parseDouble(part.trim()));
                    } catch(NumberFormatException e){
                        // 无论是因 dependent_var 还是独立变量，只返回相同错误信息以符合测试要求
                        throw new IllegalArgumentException("Dependent variable contains non-numeric data.");
                    }
                }
            }
            dataMap.put(key, values);
        }

        // 如果解析结果为空，则输入不是字典
        if(dataMap.isEmpty()){
            throw new IllegalArgumentException("Data must be a dictionary.");
        }

        // 检查是否包含依赖变量
        if(!dataMap.containsKey(dependent_var)){
            throw new NoSuchElementException("KeyError: " + dependent_var);
        }
        List<Double> yList = dataMap.get(dependent_var);
        // 检查依赖变量中数据是否足够
        if(yList.size() < 2){
            throw new IllegalArgumentException("Insufficient data for regression.");
        }
        // 检查依赖变量数据是否全为数字（解析时已保证）
        for(Double d : yList){
            if(d == null){
                throw new IllegalArgumentException("Dependent variable contains non-numeric data.");
            }
        }

        // 从字符串中解析自变量列表（以逗号分隔）
        String[] indepVars = independent_vars_str.split(",");
        List<String> indepVarList = new ArrayList<>();
        for(String s : indepVars){
            indepVarList.add(s.trim());
        }

        int n = yList.size(); // 样本数量
        int p = indepVarList.size(); // 自变量个数

        // 构造X矩阵，第一列是常数项
        double[][] X = new double[n][p+1];
        double[] y = new double[n];
        for(int i=0; i<n; i++){
            X[i][0] = 1.0; // 常数项
            for(int j=0; j<p; j++){
                String key = indepVarList.get(j);
                if(!dataMap.containsKey(key)){
                    throw new NoSuchElementException("KeyError: " + key);
                }
                List<Double> col = dataMap.get(key);
                X[i][j+1] = col.get(i); // 假设每个列表长度一致
            }
            y[i] = yList.get(i);
        }

        // 调用Apache Commons Math实现OLS回归
        OLSMultipleLinearRegression regression = new OLSMultipleLinearRegression();
        // 使用我们手动添加的常数项，因此设置不自动截距
        regression.setNoIntercept(false);
        regression.newSampleData(y, X);
        double rSquared = regression.calculateRSquared();
        double[] params = regression.estimateRegressionParameters();

        // 构造回归系数字典，第一个参数为常数项
        Map<String, Double> coefMap = new LinkedHashMap<>();
        coefMap.put("const", params[0]);
        for(int j=0; j<p; j++){
            coefMap.put(indepVarList.get(j), params[j+1]);
        }

        // 返回格式与python中一致
        return "R-squared: " + rSquared + ", Coefficients: " + coefMap.toString();
    }
}